File: Sliding_Window/LC_76.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.28

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        count_s, count_t  = collections.defaultdict(int), collections.defaultdict(int)
        for ch in t:
            count_t[ch] += 1
        need = len(count_t)
        start, end, ans_len = -1, -1, float('inf')
        have = 0
        l, r = 0, 0
        for r in range(len(s)):
            count_s[s[r]] += 1
            if s[r] in count_t and count_s[s[r]] == count_t[s[r]]:
                have += 1
            # shrink the left side if have == need
            while have == need:
                # update the answer first
                lens = r - l + 1
                if lens < ans_len:
                    ans_len = lens
                    start, end = l, r
                # move left pointers
                count_s[s[l]] -= 1
                if s[l] in count_t and count_s[s[l]] < count_t[s[l]]:
                    have -= 1
                l += 1
        #print(start, end, ans_len)
        return s[start:end+1] if ans_len != float('inf') else ""

File: Sliding_Window/LC_76.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.28

class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> s_count, t_count;
        for (char ch : t){
            t_count[ch]++;
        }
        int need = t_count.size(); // how many unique chars
        int have = 0;
        int start = 0, ans_len = INT_MAX;
        int l = 0;
        for (int r = 0; r < s.size(); r++){
            s_count[s[r]]++;
            if (t_count.count(s[r]) && s_count[s[r]] == t_count[s[r]]){
                have += 1;
            }
            // keep shrinking the left pointer to make the substring minimal
            while (have == need){
                // update the answer first
                int cur_len = r - l + 1;
                if (cur_len < ans_len){
                    ans_len = cur_len;
                    start = l;
                }
                s_count[s[l]]--;
                if (t_count.count(s[l]) && s_count[s[l]] < t_count[s[l]]){
                have -= 1;
                }
                l++; 
            }
        }
        if (ans_len == INT_MAX){
            return "";
        }
        return s.substr(start, ans_len);

    }
};